controller/src/main/java/io/pravega/controller/store/stream/ZKStreamMetadataStore.java

```diff
@@ -70,6 +97,82 @@ ZKStream newStream(final String scope, final String name) {
+            } else {
+                future = CompletableFuture.completedFuture(next);
+            }
+        }
+        return future;
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> refreshRangeIfNeeded() {
+        CompletableFuture<Void> refreshFuture;
+        synchronized (lock) {
+            // Ensure that only one background refresh is happening. For this we will reference the future in refreshFutureRef
+            // If reference future ref is not null, we will return the reference to that future.
+            // It is set to null when refresh completes.
+            refreshFuture = this.refreshFutureRef;
+            if (this.refreshFutureRef == null) {
+                // no ongoing refresh, check if refresh is still needed
+                if (counter.get().compareTo(limit.get()) >= 0) {
+                    log.info("refreshing counter range. Current counter is {}. Current limit is {}", counter.get(), limit.get());
+
+                    // Need to refresh counter and limit. Start a new refresh future. We are under lock so no other
+                    // concurrent thread can start the refresh future.
+                    refreshFutureRef = getRefreshFuture()
+                            .exceptionally(e -> {
+                                // if any exception is thrown here, we would want to reset refresh future so that it can be retried.
+                                synchronized (lock) {
+                                    refreshFutureRef = null;
+                                }
+                                log.warn("Exception thrown while trying to refresh transaction counter range", e);
+                                throw new CompletionException(e);
+                            });
+                    // Note: refreshFutureRef is reset to null under the lock, and since we have the lock in this thread
+                    // until we release it, refresh future ref cannot be reset to null. So we will always return a non-null
+                    // future from here.
+                    refreshFuture = refreshFutureRef;
+                } else {
+                    // nothing to do
+                    refreshFuture = CompletableFuture.completedFuture(null);
+                }
+            }
+        }
+        return refreshFuture;
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> getRefreshFuture() {
+        return storeHelper.createZNodeIfNotExist(ZKStoreHelper.COUNTER_PATH, BigLong.ZERO.toBytes())
+                .thenCompose(v -> storeHelper.getData(ZKStoreHelper.COUNTER_PATH)
+                        .thenCompose(data -> {
+                            BigLong previous = BigLong.fromBytes(data.getData());
+                            BigLong nextLimit = BigLong.add(previous, COUNTER_RANGE);
+                            return storeHelper.setData(ZKStoreHelper.COUNTER_PATH, new Data<>(nextLimit.toBytes(), data.getVersion()))
```

Comment:
I'm not sure I understand how you are handling the case of a bad version error in this conditional set data. Don't we have to try again?

Reply:
The createTransaction method in streamTransactionMetadataTasks has an uber retryBlock which will kick in and retry on retryable exceptions.